src/main/java/com/mongodb/hibernate/jdbc/MongoStatement.java

```diff
@@ -228,21 +279,379 @@ void checkClosed() throws SQLException {
+            checkStatementFields(deleteStatement, commandDescription, SUPPORTED_DELETE_STATEMENT_FIELDS);
+            var isSingleDelete = deleteStatement.getNumber("limit").intValue() == 1;
+            var filter = deleteStatement.getDocument("q");
+
+            if (isSingleDelete) {
+                return new DeleteOneModel<>(filter);
+            }
+            return new DeleteManyModel<>(filter);
+        }
+
+        private static void checkStatementFields(
+                BsonDocument statement, CommandDescription commandDescription, Set<String> supportedStatementFields)
+                throws SQLFeatureNotSupportedException {
+            checkFields(
+                    commandDescription,
+                    UNSUPPORTED_MESSAGE_STATEMENT_FIELD,
+                    supportedStatementFields,
+                    statement.keySet().iterator());
+        }
+
+        private static void checkCommandFields(
+                BsonDocument command, CommandDescription commandDescription, Set<String> supportedCommandFields)
+                throws SQLFeatureNotSupportedException {
+            var iterator = command.keySet().iterator();
+            iterator.next(); // skip the command name
+            checkFields(commandDescription, UNSUPPORTED_MESSAGE_COMMAND_FIELD, supportedCommandFields, iterator);
+        }
+
+        private static void checkFields(
+                CommandDescription commandDescription,
+                String exceptionMessage,
+                Set<String> supportedCommandFields,
+                Iterator<String> fieldNameIterator)
+                throws SQLFeatureNotSupportedException {
+            while (fieldNameIterator.hasNext()) {
+                var field = fieldNameIterator.next();
+                if (!supportedCommandFields.contains(field)) {
+                    throw new SQLFeatureNotSupportedException(
+                            exceptionMessage.formatted(commandDescription.getCommandName(), field));
+                }
+            }
+        }
+    }
+
+    /** Maps write model indices to their corresponding command indices in batch of commands. */
+    private static class WriteModelsToCommandMapper {
+        /** The cumulative counts of write models for each command in the batch (prefix sum). */
+        private final int[] cumulativeCounts;
+
+        private int cumulativeCountIndex;
+
+        private WriteModelsToCommandMapper(int commandCount) {
+            this.cumulativeCounts = new int[commandCount];
+            this.cumulativeCountIndex = 0;
+        }
+
+        private void add(int cumulativeWriteModelCount) {
+            assertFalse(cumulativeCountIndex >= cumulativeCounts.length);
+            cumulativeCounts[cumulativeCountIndex++] = cumulativeWriteModelCount;
+        }
+
+        private int findCommandIndex(int writeModelIndex) {
+            assertTrue(cumulativeCountIndex == cumulativeCounts.length);
+            int lo = 0, hi = cumulativeCounts.length;
```

Comment:
We have never used this syntax in any of our codebases, as far as I know. Do we really want to start using it? I would not say it improves readability, but it does reduce the number of lines.